compiler/rustc_resolve/src/imports.rs

```diff
@@ -828,16 +829,33 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
                     Err(Undetermined) => indeterminate_count += 1,
                     // Don't update the resolution, because it was never added.
                     Err(Determined) if target.name == kw::Underscore => {}
-                    Ok(binding) if binding.is_importable() => {
+                    Ok(binding)
+                        if binding.is_importable()
+                            || binding.is_assoc_const_or_fn()
+                                && this.tcx.features().import_trait_associated_functions() =>
+                    {
```

Comment:
I think `is_importable` should just stop rejecting assoc const and fn. Instead, within this match arm, emit the feature error unconditionally if it is either of those two (and the feature wasn't enabled). That's how we handle feature gating where possible/feasible: just go through as if the feature gate was already enabled, but emit an error if it wasn't.

Reply:
Oof. I did not consider the implications of this. Now we get suggestions to import the trait even without the feature gate, but actually following the suggestion will error due to the feature gate missing. So your original idea was right (the just acting as if the feature gate were enabled but emitting the feature error is still good tho). Please switch the body of `is_importable` back (but add a `FIXME(import_trait_associated_functions)` to it, so after stabilization we'll make this change).

sorry for the back and forth

Reply:
> just go through as if the feature gate was already enabled, but emit an error if it wasn't.

I think your original suggestion still makes sense, but we still need to avoid emitting suggestions that lead to errors due to the feature gate missing.